make sure we detect journal rotation even on inotify q overrun

poettering opened the pull request:
Plus some other fixes.

Fixes: #8198.

portante commented on src/basic/fs-util.c:
```diff
@@ -967,3 +967,33 @@ int unlinkat_deallocate(int fd, const char *name, int flags) {
+        if (!path_is_absolute(path))
+                return -EINVAL;
+
+        dn = dirname_malloc(path);
```
What entity calls `free()` on `dn`?

vcaputo replied:
@portante note the _cleanup_free_ qualifier on dn's type and how it's initialized to NULL.  This is magic that frees dn when the function returns if dn is != NULL.

vcaputo commented:
FWIW I gave this a quick casual pass and it seemed ok, but nothing thorough enough to give a solid lgtm.

keszybz commented on src/journal/journal-file.c:
```diff
@@ -3312,7 +3314,11 @@ int journal_file_open(
+                 * fail quickly than block for long. For regular files O_NONBLOCK has no effect, hence it doesn't hurt
+                 * in that case. */
+
+                f->fd = open(f->path, f->flags|O_CLOEXEC|O_NONBLOCK, f->mode);
```
Hmm, open(2) says:
> Note that this flag has no effect for regular files  and  block  devices

so the comment seems wrong, but what's worse it also says
> Since  O_NONBLOCK semantics might eventually be implemented, applications should not depend upon blocking behavior when specifying this flag for regular files and block devices.

which means that (potentially), our code might stop working in the future. Since this is in library code, that could even be potentially running on non-linux systems, we shouldn't depend on the implementation details of the kernel.

poettering replied:
I figure i should say char device, not block device... any maybe also list fifos or so...

well, the stuff is pretty linux specific anyway. if somebody indeed ports that to other OSes, then they need to investigate this kind of stuff anyway... I am pretty sure it's highly unlikely O_NONBLOCK will ever be implemented for regular files on Linux, given that usage like the above is seen quite often.

The primary reason I added this is that we iterate trhough a directory tree and open all potentially useful journal files we see there. i.e. it's not that the knowledgeable admin would never point us to weird device files and such, it's that we go out on our own to find stuff to open and open it, and that makes things dangerous, and something we should protect us from...

keszybz replied:
OK.

keszybz commented on src/journal/sd-journal.c:
```diff
@@ -1254,81 +1264,113 @@ static const char *skip_slash(const char *p) {
-                /* If there's a top-level fd defined, open the file relative to this now. (Make the path relative,
-                 * explicitly, since otherwise openat() ignores the first argument.) */
+        if (fstat(fd, &st) < 0) {
+                r = log_debug_errno(errno, "Failed to open file '%s': %m", path);
```
Failed to fstat...

keszybz commented on src/journal/sd-journal.c:
```diff
@@ -1254,81 +1264,113 @@ static const char *skip_slash(const char *p) {
+                goto finish;
+        }
+        if (S_ISDIR(st.st_mode)) {
+                log_debug("Uh, file '%s' is a directory? Refusing.", path);
```
I'd prefer `"Cannot add journal file \"%s\": %m"`...

keszybz commented on src/journal/sd-journal.c:
```diff
@@ -1254,81 +1264,113 @@ static const char *skip_slash(const char *p) {
+                goto finish;
+        }
+        if (!S_ISREG(st.st_mode)) {
+                log_debug("Uh, file '%s' is not a regular file? Refusing.", path);
```
I'd prefer "Cannot add journal file \"%s\": not a regular file"...

poettering replied:
a later commit changes these two messages anyway

keszybz replied:
I still see it in the final diff.

keszybz commented on src/journal/sd-journal.c:
```diff
@@ -1254,81 +1264,113 @@ static const char *skip_slash(const char *p) {
+
+                        /* We already track this file, under the same path and with the same device/inode numbers, it's hence
+                         * really the same. Mark this file as seen in this generation. This is used to GC old files in
+                         * process_q_overflow() to detect journal files that are still and discern them from those who are
```
that are still *what*?

msekletar commented:
@poettering thank you for working on this! FWIW, I force pushed my branch in #8204 with the code that I wrote yesterday. As explained in commit message it doesn't work and I'd like to understand why.

keszybz commented:
OK, so it's tested and all significant issues are resolved. @poettering can you fix the typos?

poettering commented:
I force pushed a new version. I fixed the typos @keszybz found, plus some other things:

1. I changed the O_NONBLOCK code to turn off O_NONBLOCK again after we opened the file and decided to stick to it. That means we'll only do an async open, but nothing else. I kind hope to address @keszybz's points this way a bit: we do get the behaviour I was looking for (i.e. non blocking open() of weird devices and fifos) but at the same time we don't actually use O_NONBLOCK afterwards, hence in the unlikely case the kernel defines an O_NONBLOCK mode for regular files after all one day we should still be fine.

2. I changed the code that fakes the "creation"/"birth" time of a file with xattrs to also consult the actual "btime", as current Linux kernels finally expose this information where available using statx. The code will still read the old xattr though to make the best of the situation as there are kernels/file systems that support btime, and others which support xattr, as well as those supporting both (and those supporting either). In the interest of best compatibility we'll hence read both and use the older value. [the btime/crtime is used for figuring out which files to delete first when doing journal or coredump vacuuming]

3. i added a commit fixing a log message typo elsewhere

poettering commented:
> @poettering thank you for working on this! FWIW, I force pushed my branch in #8204 with the code that I wrote yesterday. As explained in commit message it doesn't work and I'd like to understand why.

You mean, your code in #8204 doesn't work? Hmm, quite frankly: I have no idea. Reading through it made a lot of sense to me, I don't see why it wouldn't work. It essentially does the same as my patch actually, hence I am a bit puzzled...

msekletar commented:
Exactly, I don't know either why it doesn't work. Nevertheless, your version works so I don't really care anymore.
